// piezo-server/libs/lists.js
/*!
 * Module: Module
 *
 * @author Andjey Guzhovskiy, <me.the.ascii@gmail.com>
 * @copyright (c) 2013 Andjey Guzhovskiy
 * @licence CLOSED
 * @version 0.0.1
 */

var mongo = require( 'mongodb' ),
    colors = require( 'colors' ),
    client,
    collection,
    config = {
        // connection
        port: 27017,
        host: 'localhost',
        write: 1,
        // data
        database: 'test',
        collection: 'cards'
    };


// Base

exports.app = null;
exports.configure =
    function( cfg ) {
        var cfg = cfg || {},
            defaults = cfg.defaults || {};

        // config
        if ( cfg.host
            || defaults.mongodb.host )
            config.host = cfg.host || defaults.mongodb.host;
        if ( cfg.port
            || defaults.mongodb.port )
            config.port = cfg.port || defaults.mongodb.port;
        if ( cfg.database
            || defaults.mongodb.database )
            config.database = cfg.database || defaults.mongodb.database;
        if ( cfg.collection
            || defaults.mongodb.collection )
            config.collection = cfg.collection || defaults.mongodb.collection;
    };

exports.init =
    function( libs, callback ) {

        // database
        client = new mongo.Db(
            config.database,
            new mongo.Server( config.host, config.port, { auto_reconnect: true }),
            { w: config.write }
        );

        client.on( 'error', function( err ) {
            console.log( 'catch:'.bold.red, err );
        });

        // connect db
        client.open( function( err, _client ) {
            if ( err ) {
                console.log( 'Error: Auth connect database:'.bold.red, err );
                process.exit( 2 );
            }
            client.collection( config.collection,
                function( err, coll ) {
                    if ( err ) {
                        console.log( 'Error: Auth database collection:'.bold.red, err );
                        process.exit( 2 );
                    }
                    collection = coll;

                    // success
                    console.log( 'Lists storage:'.magenta, 'mongo at', config.host +':'+ config.port +'/'+ config.database );
                    if ( callback ) callback();
                });
        });
    };


// List ( collection )

exports.createList =
    function( list, callback ) {};
exports.getList =
    function( list, callback ) {};
exports.updateList =
    function( id, list, callback ) {};
exports.removeList =
    function( id, callback ){};


// Item ( model )

exports.createItem =
    function( item, callback ) {};
exports.getItem =
    function( item, callback ) {};
exports.updateItem =
    function( id, item, callback ) {};
exports.removeItem =
    function( id, callback ) {};
