// piezo-server/libs/mail.js
/*!
 * Module: Module
 *
 * @author Andjey Guzhovskiy, <me.the.ascii@gmail.com>
 * @copyright (c) 2013 Andjey Guzhovskiy
 * @licence CLOSED
 * @version 0.0.1
 */

var util = require( 'util' ),
    nodemailer = require( 'nodemailer' ),
    fs = require( 'fs' ),
    check = require( 'validator' ).validators,
    filters = require( 'validator' ).sanitize,
    //
    client,
    config = {
        // connection
        proto: 'smtp',
        service: 'Mailgun',
        user: 'api',
        password: 'key-'
    };

// todo: queue

// Boot

// Base

exports.app = nodemailer;
exports.configure =
    function( cfg ) {
        var cfg = cfg || {};

        // config
        if ( cfg.user )
            config.user = cfg.user;
        if ( cfg.password )
            config.password = cfg.password;
    };

exports.init =
    function( libs, callback ) {

        // mailer
        client = nodemailer.createTransport(
            config.proto.toUpperCase(),
            {
                service: config.service,
                auth: {
                    user: config.user,
                    pass: config.password
                }
            });

        console.log( 'Mail system run:'.magenta, config.service );
        callback && callback();
    };


// API

exports.send =
    function( mailOptions, callback ) {
        var opts = options( mailOptions );

        // params
        if ( !client )
            return callback( new Error( 'Mail client is not ready' ));
        if ( !opts )
            return callback( new Error( 'Bad mail options:', opts ));

        console.log( 'send mail:', opts );

        // send
        client.sendMail( opts,
            function(err, res) {
                if ( err ) {
                    console.log( 'Mail send error:', err );
                    callback && callback( err );
                } else {
                    console.log( 'Mail sent'.green, opts.to );
                    callback && callback( null, res );
                }
            });
//        callback && callback();
    };

exports.options = options;
function options( opt ) {
    var mail = {},
        opt = opt || {};

    mail.from = opt.from;
    mail.to = opt.to;
    mail.subject = opt.subject;
    mail.html = opt.body;
    //mail.attachments = opt.attachments;

    // checks
    if ( !mail.from )
        // || !check.isEmail( mail.form ))
        return console.log( 'Bad mail from:', mail.from );
    if ( !mail.to )
        // || !check.isEmail( mail.to ))
        return console.log( 'Bad mail to:', mail.to );
    if ( !mail.subject )
        return console.log( 'No mail subject' );
    if ( !mail.html )
        return console.log( 'Empty mail body' );
//    if ( !mail.attachments
//        || !util.isArray( mail.attachments ))
//        return console.log( 'No mail attachments' );

    // todo: filters

    // valid
    return Object.keys( mail ).length
        ? mail
        : null;
}
